Fix docker cp

andergnet opened the pull request:
This pull request fixes issue #18

kolayne commented on driver.go:
```diff
@@ -273,40 +233,20 @@ func (d *DockerOnTop) Unmount(request *volume.UnmountRequest) error {
-			"manually to fix the problem", err)
+		log.Debugf("Unmounted volume %s", request.Name)
+	} else {
+		log.Debugf("Volume %s is still mounted. Indicating success without unmounting", request.Name)
```
I don't agree with the error message change here. Here, you're seemingly saying "still mounted so I'm not unmounting it", which doesn't make sense. That's because of the ambiguity between "mounted" as in "a container is running where this volume is used" and "mounted" as in "the underlying overlay filesystem is present". IMO, the previous error message was clearer:
```go
log.Debugf("Volume %s is still mounted in some other container. Indicating success without unmounting", request.Name)
```

Maybe could be phrased even better, e.g., `Volume %s is still used in another container. Indicating success without unmounting`

andergnet replied:
I like the one you proposed, it's more clear of what's happening:
197c6feb30581a079b201dab667119abb8009af1

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+	UsageCount int
+}
+
+func (d *DockerOnTop) Activate(request *volume.MountRequest, activemountsdir lockedFile) (bool, error) {
```
A comment on both `DockerOnTop.Activate` and `DockerOnTop.Activate`:

1.    Their names probably don't quite reflect what they do. If I saw the following piece of code:
      ```go
      var d DockerOnTop = ...
      d.Activate(...)
      ```
      I would most probably assume that `d.Activate` _activates_ the `d` object (whatever it could mean) because that's what the name suggests. But what it actually does is it activates _a volume_. If we had a proper volume object, it would make sense to move this method there and have `volume.Activate`, but as its a method of `DockerOnTop`, its name should probably be `ActivateVolume`

2.  In go, the case of the first letter (i.e., whether it's lowercase or uppercase) has a semantic effect: if an entity name is capitalized, it's _exported_ (i.e., public - available from other packages), otherwise it's _unexported_ (i.e., private, only available in this package). As I understand, the `.Activate` and `.Deactivate` are both intended for internal use only (a caller external to `DockerOnTop` shall not call them directly). If that's the case, their names should be changed.

3.  If I'm wrong and you intended that these functions are exported, it's a go convention to have a documentation string for every exported entity (see [Documenting go code](https://go.dev/blog/godoc) for details). You may notice that other exported methods of `DockerOnTop` don't have documentation, which is a flaw, but all of them correspond to Docker API calls, whereas the two new functions you are adding are something new.

4.  Finally, as you may notice in further comments, at some points it wasn't clear to me what these functions should do and, especially, what they should return. That's why, even if you didn't mean to export them and you'll rename them to `.activateVolume` and `.deactivateVolume` respectively, they would still benefit from a short docstring that explains what they do and what they return.

andergnet replied:
I hope this looks better now with the changes I have done:
Commit: b18043e55f229afdeb0243239d7f25d55d012173

andergnet replied:
Also commit 3a67652e35100e08b72481ca830ffc1c62d54e19 since I didn't properly rename `deactivateVolume`

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+	"github.com/docker/go-plugins-helpers/volume"
+)
+
+type ActiveMount struct {
```
You probably meant to call it `activeMount` (see point 2 of the next note, on exporting entities in go).
(note that unlike the structure, its field _should_ be exported: otherwise `json.Marshal`/`.Unmarshal` won't be able to access it, as they are not part of this package)

andergnet replied:
It's my first real program in Go, I had no idea about that, I simply copied how `VolumeInfo` was defined. I have fixed it.

kolayne reviewed (changes requested):
Good evening! Sorry, I didn't notice the PR when you opened it and then didn't check my email for almost a week... :sob: 

But here I am :)

Firstly, thank you for the PR! There are some changes that I request below. Please, see my comments and let me know if I misunderstood something that you did intentionally.

Second, a general note (too late for this PR but may be useful for your future contributions): when your PR, while addressing a single problem, can be broken down into logical sub-steps, which you take to implement the patch, many maintainers prefer that you make a commit per step, so that your changes can be reviewed step by step as well. For example, I see two logical sub-steps of this PR: first, separate the existing code between driver.go and activemount.go, next, patch the multiple-mount logic.

Btw, providing detailed commit messages, answering the what&why questions ("what did you do? why did you do that?") is also beneficial and gives the reviewer an insight into your changes before they even starts reading your code!

<br>

UPD: on this page my comments are displayed in the order I created them, but it's better if you read them along with the code, in the order they appear in the [Files changed](https://github.com/kolayne/docker-on-top/pull/20/files) section

kolayne commented on driver.go:
```diff
@@ -184,10 +183,10 @@ func (d *DockerOnTop) Mount(request *volume.MountRequest) (*volume.MountResponse
+	mount, err := d.Activate(request, activemountsdir)
+	if err != nil {
+		return nil, internalError("failed to activate the active mount:", err)
+	} else if mount {
 		lowerdir := thisVol.BaseDirPath
 		upperdir := d.upperdir(request.Name)
 		workdir := d.workdir(request.Name)
```
Read [this comment](https://github.com/kolayne/docker-on-top/pull/20/files#r1693985907) first.

<br>

Here you, unfortunately, did kind of the opposite (similarly, you reversed the original order, but it has the inverse effect in this case).

Combining the logic inside `d.Activate` and the code _below_ this comment up to and including the `syscall.Mount` (that's the code you did not modify), the overall (new) logic is as follows:

- Check if the volume is already mounted based on `activemount` files. If it's not:
- Create a new `activemount` file to indicate that I am the user of the volume \[may fail\]
- Perform the actual mount of the `overlay` \[may fail\]

Now, imagine the file creation step succeeds but the next one fails. For that container, of course, we'll report the error and it will refuse to start, but the volume ends up in a _very bad_ state. If another container tries to use that same volume, it will perform the same steps above, but in the first one it will find an activemount file from the previous container and will assume that the volume is already mounted, while it in fact is not!!

The reverse order is not very good either but slightly better: if the first container manages to mount the volume but doesn't manage to mark it as mounted, it, likewise, won't let the current container start (so that `overlay` mount remains unused), then, when the next container attempts a mount, it will run the same steps and, in the trickiest case, will just mount the same `overlay` on top of the old one and start using it (which is weird because there are two `overlay`s mounted simultaneously using the same internal directories, but should be ok, as the first overlay is never used, being shadowed by the new one).

<hr>

So,

1. First possible improvement for your code: your implementation should at least emit a scary screaming error! (see [this example](https://github.com/kolayne/docker-on-top/blob/1bfffce2ff8960b549a9d5fb0a894895163e5ca6/driver.go#L301-L309) from the old code - this error was used in a very similar situation)
2. Second, it would be much better if you could fix things such that the order of the system calls (mount/unmount call and activemount file creation/deletion call) is old (as in my code) for the `Mount` operation but is new (as in your code) for the `Unmount` operation.

In fact, I now see how it's possible to completely fix the problem in the `Mount` part, so that we never even attempt to mount `overlay` on top of anything old. Of course, this is out of scope of this PR, so I'll create a separate issue (but it still requires that you restore the old order here in `Mount`)

andergnet replied:
I have improved the way Mount function works, I keep the new order but in case the filesystem mount fails it will undo the `activemount` creation so it returns the function in a consistent state: filesystem not mounted and no activemount present.

01de48d2fe2c0cfc4fb01f1fd34247ca88a11e98

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+func (d *DockerOnTop) Activate(request *volume.MountRequest, activemountsdir lockedFile) (bool, error) {
+	var activeMount ActiveMount
```
Fixed 7043a09038fa4b8000db47c9ab3f1d4d99ea832d

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+
+	otherVolumesPresent := len(dirEntries) > 1 || dirEntries[0].Name() != request.ID
+
+	file, err := os.Open(activemountFilePath)
```
Fixed: c0d28c03e0834a7b8b2fda8039376bbf41fe25ea

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+
+	activeMount.UsageCount++
+
+	payload, _ := json.Marshal(activeMount)
```
_it's not clear to me why. json.Marshal's behavior should be much more predictable than that of json.Unmarshal_
We have control over what the Marshal-ed structure (activeMount) so we will not get suprises there. Unmarshal, on the other hand, deals with a file in the filesystem and we could have a surprise if someone changes it.

I have added a comment to indicate why it is safe to ignore the error, which basically boils down to the fact that there is no room for error when converting an structure containing a single integer to JSON.
03f2054204f32b32c670c3b866301210b7518652

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+	activeMount.UsageCount++
+
+	payload, _ := json.Marshal(activeMount)
+	err = os.WriteFile(activemountFilePath, payload, 0o666)
```
Indeed, 0764601aad29eeb7de353eb28e26a5f3be7aad89

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+		}
+		return !otherVolumesPresent, nil
+	} else {
+		payload, _ := json.Marshal(activeMount)
```
03f2054204f32b32c670c3b866301210b7518652

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+		return !otherVolumesPresent, nil
+	} else {
+		payload, _ := json.Marshal(activeMount)
+		err = os.WriteFile(activemountFilePath, payload, 0o666)
```
You give me too much credit 😉, I probably copy-pasted that from somewhere else and didn't give it a second thought. Thankfully revisions exist.
0764601aad29eeb7de353eb28e26a5f3be7aad89

andergnet commented on driver.go:
```diff
@@ -184,10 +183,10 @@ func (d *DockerOnTop) Mount(request *volume.MountRequest) (*volume.MountResponse
-	if errors.Is(readDirErr, io.EOF) {
-		// No files => no other containers are using the volume. Need to mount the overlay
-
+	mount, err := d.Activate(request, activemountsdir)
```
Changed to doMountFs also to differenciate between the two meanings of "Mount".
1985b5201fee166a4fba0d11e1a8c1026c247f02

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+
+func (d *DockerOnTop) Activate(request *volume.MountRequest, activemountsdir lockedFile) (bool, error) {
+	var activeMount ActiveMount
+	var result bool
```
Since this was an issue also in driver.go (see [comment](https://github.com/kolayne/docker-on-top/pull/20/files/bb0d01096262eea9b5f04e53851963f7f45fe453#r1694002909)), I have named it the same way everywhere: 1985b5201fee166a4fba0d11e1a8c1026c247f02

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+	} else {
+		log.Errorf("Active mount file %s exists but cannot be read.", activemountFilePath)
+		return false, fmt.Errorf("active mount file %s exists but cannot be read", activemountFilePath)
+	}
```
I have applied the suggestions as they do make good sense. As if it was intended, the case and punctuation changes they were, initially they where the same with human readable format (casing and sentence ending dots) but the linter complained about that format on the error message. That's how it ended up like this. 

c57da7724d6098d4bcb9ec34cb4a554b370576a6
f9aadb4df4833d2d81d15aee3ed93c30c52f9ba6

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+	if err != nil {
+		log.Errorf("Active mount file %s cannot be written.", activemountFilePath)
+		return false, fmt.Errorf("active mount file %s cannot be written", activemountFilePath)
+	}
```
c57da7724d6098d4bcb9ec34cb4a554b370576a6
f9aadb4df4833d2d81d15aee3ed93c30c52f9ba6

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+	if errors.Is(readDirErr, io.EOF) {
+		// If directory is empty, unmount overlay and clean up
+		log.Errorf("There are no active mount files and one was expected. Unmounting.")
+		return true, fmt.Errorf("there are no active mount files and one was expected. Unmounting")
```
c57da7724d6098d4bcb9ec34cb4a554b370576a6
f9aadb4df4833d2d81d15aee3ed93c30c52f9ba6

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+	} else {
+		log.Errorf("The active mount file %s could not be opened", activemountFilePath)
+		return false, fmt.Errorf("the active mount file %s could not be opened", activemountFilePath)
+	}
```
c57da7724d6098d4bcb9ec34cb4a554b370576a6
f9aadb4df4833d2d81d15aee3ed93c30c52f9ba6

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+		err = os.WriteFile(activemountFilePath, payload, 0o666)
+		if err != nil {
+			log.Errorf("The active mount file %s could not be updated", activemountFilePath)
+			return false, fmt.Errorf("the active mount file %s could not be updated", activemountFilePath)
```
c57da7724d6098d4bcb9ec34cb4a554b370576a6
f9aadb4df4833d2d81d15aee3ed93c30c52f9ba6

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+		// The file can exist from a previous mount when doing a docker cp on an already mounted container, no need to mount the filesystem again
+		payload, _ := io.ReadAll(file)
+		json.Unmarshal(payload, &activeMount)
+		file.Close()
```
Fixed: f6634d9b5dabdaaa5322ea9b4f000a8fed5bbcc0

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+	if err == nil {
+		payload, _ := io.ReadAll(file)
+		json.Unmarshal(payload, &activeMount)
+		file.Close()
```
Fixed: f6634d9b5dabdaaa5322ea9b4f000a8fed5bbcc0

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,110 @@
+		err := os.Remove(activemountFilePath)
+		if err != nil {
+			log.Errorf("The active mount file %s could not be deleted", activemountFilePath)
+			return false, fmt.Errorf("the active mount file %s could not be deleted", activemountFilePath)
```
I'm not really convinced with the new text either, if you have a suggestion please do not hesitate to propose.
Now that the log messages are handled in driver.go, i would keep all of them as Errorf and not Criticalf for the sake of simplicity and the fact that I do not consider the errors more severe.
e441320f90110aebe797cc8e412f493e2a59cd81

andergnet commented:
@kolayne I think I have addressed all the issues you raised in the review. Could you please review it again?

I tried to be more transparent and show every step of the way but I'm not used to do it, so my apologies if you expected something different. 
I won't have access to a proper computer for a few weeks, don't worry if I do not answer I'll do it when I'm back.

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+		closeErr := file.Close()
+		if readErr != nil {
+			return false, fmt.Errorf("active mount file %s has been opened but cannot be read %w", activemountFilePath, readErr)
+		} else if closeErr != nil {
+			return false, fmt.Errorf("active mount file %s has been opened but cannot be closed %w", activemountFilePath, readErr)
+		}
```
Failing to close a file is bad enough to emit a warning but I don't think it's bad enough to abort the operation (if we read the contents successfully, there's no problem to continue). I'd do something like:
```go
if err := file.Close(); err != nil {
    log.Warningf("Failed to close active mount file: %v", err)
}
```

andergnet replied:
I have replaced everything with a os.ReadFile

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+
+	var activeMountInfo activeMount
+	activemountFilePath := d.activemountsdir(requestName) + requestId
+	file, err := os.Open(activemountFilePath)
+
+	if err == nil {
+		// The file can exist from a previous mount when doing a docker cp on an already mounted container, no need to mount the filesystem again
+		payload, readErr := io.ReadAll(file)
+		closeErr := file.Close()
```
I just noticed we open the file and then on success just read all contents of it and close it. Looks just like [`os.ReadFile`](https://pkg.go.dev/os#ReadFile). Can we use it here?

Fun fact: Go's standard library [ignores](https://github.com/golang/go/blob/86bec1ec198f2720c83bd232a72b800b4ea5a9f6/src/os/file.go#L792) the error returned by `file.Close()` inside `os.ReadFile`

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+		} else if closeErr != nil {
+			return false, fmt.Errorf("active mount file %s has been opened but cannot be closed %w", activemountFilePath, readErr)
+		}
+		unmarshalErr := json.Unmarshal(payload, &activeMountInfo)
+		if unmarshalErr != nil {
```
No big deal but looks a bit too verbose to me. Could be
```go
if err := json.Unmarshal(...); err != nil {
    do stuff
}
```

andergnet replied:
If you don't have strong feelings about it I'll keep it the way it is with a single instruction per line, I find it better specially when debugging.

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+// Parameters:
+//
+//	requestName: Name of the volume to be mounted
+//	requestID: ID of the container requesting the mount
```
Factual error: `reuqestID` is not a container id (which you see, e.g., in `docker container ls`). In fact, it is some id generated for the pair {container, volume}. The doc doesn't necessarily need to be this specific, but it should not lie either.

I would call it either "a container-volume-unique ID" or, more elaborate, "an ID unique among volumes and containers but the same value if a volume is mounted for a container twice". Maybe can be phrased better.

Please, also fix the same thing in `deactivateVolume`

andergnet replied:
Done. I thought I checked it, but apparently not.

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+//
+// Return:
+//
+//	doUnmountFs: Returns true if there are not other usages of this volume and the filesystem can be unmounted.
+//	err: If the function encountered an error, the error itself, nil if everything went right.
```
```suggestion
//	doUnmountFs: `true` if there are no other usages of this volume and the filesystem shall be unmounted
//		by the caller. **Must be respected even when an error is returned**!
//	err: An error, if encountered, `nil` otherwise.
```

andergnet replied:
I have adapted your message because I have changed the behavior and now it never returns true when there is an error.

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+//
+// Return:
+//
+//	doMountFs: Returns true if the request requires the filesystem to be mounted, false if not.
+//	err: If the function encountered an error, the error itself, nil if everything went right.
```
```suggestion
//	doMountFs: `true` if the caller shall mount the filesystem, `false` otherwise.
//		If an error is returned, `doMountFs` is always `false`.
//	err: An error, if encountered, `nil` otherwise.
```

andergnet replied:
Done

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+		// If directory is empty, unmount overlay and clean up
+		return true, fmt.Errorf("there are no active mount files and one was expected. Unmounting")
+	} else if readDirErr != nil {
+		return false, fmt.Errorf("failed to list activemounts/ %v", readDirErr)
```
```suggestion
		return false, fmt.Errorf("failed to list activemounts/ : %w", readDirErr)
```

andergnet replied:
Done

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+	var activeMountInfo activeMount
+
+	activemountFilePath := d.activemountsdir(requestName) + requestId
+	file, err := os.Open(activemountFilePath)
+
+	if err == nil {
+		payload, readErr := io.ReadAll(file)
+		closeErr := file.Close()
```
Same note as above: `os.ReadFile`

andergnet replied:
Done

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+	if activeMountInfo.UsageCount == 0 {
+		err := os.Remove(activemountFilePath)
+		if err != nil {
+			return false, fmt.Errorf("the active mount file %s could not be deleted %w, the filesystem won't be unmounted", activemountFilePath, err)
```
For consistency with an earlier error, indicating action ("unmount<i>ing</i>" / "refus<i>ing</i> to unmount")

```suggestion
			return false, fmt.Errorf("the active mount file %s could not be deleted: %w. Refusing to unmount", activemountFilePath, err)
```

andergnet replied:
Instead of changing all messages saying "the filesystem won't be unmounted" I changed the first one that said "Unmount" to "the filesystem will be unmounted (which now is a warning).

kolayne commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+		payload, _ := json.Marshal(activeMountInfo)
+		err = os.WriteFile(activemountFilePath, payload, 0o644)
+		if err != nil {
+			return false, fmt.Errorf("the active mount file %s could not be updated %w, the filesystem won't be unmounted", activemountFilePath, err)
```
```suggestion
			return false, fmt.Errorf("the active mount file %s could not be written: %w. Refusing to unmount", activemountFilePath, err)
```

andergnet replied:
same as previous.

kolayne commented on driver.go:
```diff
@@ -184,10 +183,11 @@ func (d *DockerOnTop) Mount(request *volume.MountRequest) (*volume.MountResponse
-
+	doMountFs, err := d.activateVolume(request.Name, request.ID, activemountsdir)
+	if err != nil {
+		log.Errorf("Error while activating the filesystem mount: %w", err)
```
`%w` is a feature of `fmt.Errorf` only, when you want to create an error _object_. As `log.Errorf` just prints something, rather than creating an object, it uses `fmt.Sprintf` under the hood, which will either silently interpret `%w` as `%v`, or will display some warning or something. In any case,
```suggestion
		log.Errorf("Error while activating the filesystem mount: %v", err)
```

andergnet replied:
Done this one and another one in line 250.

kolayne commented on driver.go:
```diff
@@ -201,57 +201,29 @@ func (d *DockerOnTop) Mount(request *volume.MountRequest) (*volume.MountResponse
-		} else if err != nil {
-			log.Errorf("Failed to mount overlay for volume %s: %v", request.Name, err)
-			return nil, internalError("failed to mount overlay", err)
+		if err != nil {
+			// The filesystem could not be mounted so undo the activateVolume call so it does not appear as if
+			// we are using a volume that we couln't mount. We can ignore the doUnmountFs because we know the volume
+			// is not mounted.
```
TODO for me: I will look into it more carefully some time later

kolayne commented on driver.go:
```diff
@@ -273,40 +245,21 @@ func (d *DockerOnTop) Unmount(request *volume.UnmountRequest) error {
+	if err != nil {
+		log.Errorf("Error while activating the filesystem mount: %w", err)
+		return internalError("failed to deactivate the active mount:", err)
+	} else if doUnmountFs {
```
Oh-oh! From what I can see in the code of `.deactivateVolume`, it seems like you want this code to take actions for `doUnmountFs == true` even when `err != nil`! Am I right?

andergnet replied:
You are right, I did not notice there was a case returning true + error. I have changed it to log a warning instead and return true + nil

kolayne reviewed (changes requested):
Nice! Here are some more things to fix

andergnet commented on activemount.go:
```diff
@@ -0,0 +1,152 @@
+		// The directory is empty, mount the filesystem
+		doMountFs = true
+	} else {
+		return false, fmt.Errorf("failed to list activemounts/ %v", readDirErr)
```
Done.

kolayne commented:
> I'm back!

Top!


> I think I have addressed all the issues you raised, please review again.


Hopefully, I'll be able to by the end of the week

> I know it's all in a single commit, sorry, it's an old habit, and by the time I noticed I had most of the changes already done.

No problem. This mostly matters at the beginning when I'm trying to understand what's the patch about in general. When the comments are addressed, I don't look into them one by one anyway.

kolayne commented:
Hey, @andergnet!

I've found myself pretty busy with other stuff... So I don't really know when I'm going to get to merge this. That's somewhat untrivial, as I want to first merge some tests checking for some edge-case behaviors, and then merge your PR, making sure those tests pass (right now there's one scenario that is broken in master but, hopefully, with your PR, we can fix even that).

Yesterday I've made a step forward towards that, but I'm not quite there yet, and, given the experience of the last two months, I don't dare to try to estimate when I will have time to finish this.

But! I want to let you know that
1. Most likely, I won't request any more changes from you seem to have already done the core thing, and should there be minor issues, I can fix them myself;
2. I appreciate your attention to the project in general, and, in particular, the contribution with both the bug report and this workaround patch; and
3. Of course, sooner or later, this will definitely be merged!

andergnet commented:
Hi @kolayne 
Glad to hear back from you, I was wondering what happened to you since you disappeared after being quite engaged in the topic, no complains, I just hope you are doing alright (I have a job a two daughters, I understand being busy for side projects like this).

About the MR being delayed, no problem on my side, it's your project, your rules ;-). Just let you know that I have been using docker-on-top in the last few months with the patch and I haven't run into any issues.

Till next time :-)

Commit:
Fix `docker cp` on a mounted volume

This is a work around docker's bug, see moby/moby#47964.

Resolves #18

Co-authored-by: Nikolai Nechaev <Nikolay_Nechaev@mail.ru>

kolayne commented:
Hey, @andergnet, I have finally merged your changes!

As you may see, there was some code that I merged into master before your fix, then I rebased and restructured your code and merged it. FYI, the changes that I made to your patch are mostly cosmetic, except for one key amendment: the volume activation steps (mounting an overlay and creating an active mount file) and deactivation steps (removing the file and unmounting the overlay) are supposed to happen in the opposite order. So, I moved both actions into the functions you introduced, and added a note in driver.go to explain this order.

Anyway, thank you for submitting this fix (and inspiring another fix that I committed last week). By the way, hello from Korea - 안녕하세요!

andergnet commented:
환영합니다, nice to see you around!
And I see you have come with presents ;-) The changes look good and the container management looks more solid now (even if it wasn't looking that bad before). I have already merged your changes into my branch (which I'm pointing to because it already had the fix for the cp) so I start using them.

Thanks again for the effort, it's very appreciated.
